third_party/pcsc-lite/naclport/server_clients_management/src/client_request_processor.cc

```diff
@@ -215,6 +218,58 @@ void PcscLiteClientRequestProcessor::AsyncProcessRequest(
       .detach();
 }
 
+PcscLiteClientRequestProcessor::ScopedConcurrencyGuard::ScopedConcurrencyGuard(
+    const std::string& function_name,
+    SCARDCONTEXT s_card_context,
+    SCARDHANDLE s_card_handle,
+    PcscLiteClientRequestProcessor& owner)
+    : owner_(owner),
+      function_name_(function_name),
+      s_card_context_(s_card_context) {
+  s_card_context_ =
+      s_card_handle
+          ? owner_.s_card_handles_registry_.FindContextByHandle(s_card_handle)
+          : s_card_context;
+  if (!s_card_context_) {
+    return;
+  }
+
+  std::string concurrent_functions_dump;
+  {
+    const std::unique_lock<std::mutex> lock(owner_.mutex_);
+    std::multiset<std::string>& running_functions =
+        owner_.context_to_running_functions_[s_card_context_];
+    concurrent_functions_dump =
+        JoinStrings(running_functions, /*separator=*/", ");
+    running_functions.insert(function_name_);
+  }
+
+  if (!concurrent_functions_dump.empty()) {
+    GOOGLE_SMART_CARD_LOG_WARNING
+        << owner_.logging_prefix_
+        << "Client violates threading: concurrent calls of " << function_name_
+        << ", " << concurrent_functions_dump
+        << " against the same SCARDCONTEXT. Future releases of Smart Card "
+           "Connector will forbid this: every call referring to some "
+           "SCARDCONTEXT must come after the previous one completed.";
+  }
+}
+
+PcscLiteClientRequestProcessor::ScopedConcurrencyGuard::
+    ~ScopedConcurrencyGuard() {
+  if (!s_card_context_) {
+    return;
+  }
+  const std::unique_lock<std::mutex> lock(owner_.mutex_);
+  auto iter = owner_.context_to_running_functions_.find(s_card_context_);
+  GOOGLE_SMART_CARD_CHECK(iter != owner_.context_to_running_functions_.end());
+  std::multiset<std::string>& running_functions = iter->second;
+  running_functions.erase(running_functions.find(function_name_));
+  if (running_functions.empty()) {
+    owner_.context_to_running_functions_.erase(iter);
```

Comment:
Where is this actually initialized? I can't find it

Reply:
`iter` is above if that's what you mean.

Reply:
I was referring to the multiset referenced by context_to_running_functions_, but on further review initialization happens in line 236